### ISSUE TITLE
Cannot Use CUDA Memory as local buffer to ucp_put_nbx()

### ISSUE THREAD

**Issue by Seth5141** [openucx/ucx#6530]
### Describe the bug
When attempting to use CUDA allocated memory (Allocated with CuMemAlloc) as the local buffer to ucp_put_nbx, I get memory access errors. Please see below for a detailed explanation of my configuration / reproduction. But my main question is: Do the ucp_put/get_nbx operations support GPU memory in the local buffer? If so, what configuration might I be missing out on that would cause the observed behavior?

Below is the output from my program (multiprocess with 1 process each on two different nodes). The prints in the output below correspond to the starred configuration steps in the "Steps to Reproduce" section.
~~~
1: Registering memory at 0x7fc708000000 with length 1610612736
1: Pointer to local memory is 140492857183232, 0x7fc70ad88c00
1: Calling rma operation with local ptr 0x7fc70ad88c00 and remote ptr 0x7f52bba08e00 and length 4 on ep 0x7fc79007f0c0
1: [gc02:45938:0:46021] Caught signal 11 (Segmentation fault: invalid permissions for mapped object at address 0x7fc70ad88c00)
0: Registering memory at 0x7f52b8000000 with length 1610612736
0: Pointer to local memory is 139993298799616, 0x7f52bad88c00
0: Calling rma operation with local ptr 0x7f52bad88c00 and remote ptr 0x7fc70ba08e00 and length 4 on ep 0x7f53500c0140
0: [gc01:36748:0:36830] Caught signal 11 (Segmentation fault: invalid permissions for mapped object at address 0x7f52bad88c00)
~~~

Note, when I use CPU memory as the local address, but GPU memory as the remote address, everything works fine.

### Steps to Reproduce
- UCX version 1.10.0

I have confirmed the following:

- I have compiled in device memory support (See configure below)
- I have verified that the UCS_MEMORY_TYPE_CUDA flag is set when I call ucp_context_query
~~~
    ucp_context_attr_t ctx_attr;
    ctx_attr.field_mask = UCP_ATTR_FIELD_MEMORY_TYPES;
    ucs_rc = ucp_context_query(ucx_state->library_context, &ctx_attr);
    if (ucs_rc != UCS_OK) {
        ERROR_JMP(status, NVSHMEMX_ERROR_INTERNAL, error,
                  "Failed to initialize ucp for UCX.\n");
    }
    if (ctx_attr.memory_types & UCS_MEMORY_TYPE_CUDA) {
        **fprintf(stderr, "Support memory from CUDA.\n");**
    }
~~~
- I have registered the allocated memory with ucp_mem_map with the following arguments:
~~~
    **fprintf(stderr, "Registering memory at %p with length %lu\n", buf, length);**
    params.field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
                        UCP_MEM_MAP_PARAM_FIELD_LENGTH  |
                        UCP_MEM_MAP_PARAM_FIELD_FLAGS   |
                        UCP_MEM_MAP_PARAM_FIELD_PROT    |
                        UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE;
    params.flags = 0;
    params.address = buf;
    params.length = length;
    params.prot = UCP_MEM_MAP_PROT_LOCAL_READ | UCP_MEM_MAP_PROT_LOCAL_WRITE | 
    UCP_MEM_MAP_PROT_REMOTE_READ | UCP_MEM_MAP_PROT_REMOTE_WRITE;
    params.memory_type = UCS_MEMORY_TYPE_CUDA;
    ucs_rc = ucp_mem_map(ucx_state->library_context, &params, &handle->mem_handle);
~~~
- I am initiating the send with the proper flags:
~~~
    param.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK | UCP_OP_ATTR_FIELD_MEMORY_TYPE;
    param.cb.send = ucx_send_request_cb;
    param.cb.recv = ucx_recv_amo_request_cb;
    param.memory_type = UCS_MEMORY_TYPE_CUDA;
    **fprintf(stderr, "Calling rma operation with local ptr %p and remote ptr %p and length %lu on ep %p\n", local.ptr, remote.ptr, bytesdesc.elembytes * bytesdesc.nelems, ep);**
    ucs_ptr_rc = ucp_put_nbx(ep, local.ptr, bytesdesc.elembytes * bytesdesc.nelems, (uint64_t)remote.ptr, rkey, &param);
~~~

### Setup and versions
- OS version (e.g Linux distro) + CPU architecture (x86_64/aarch64/ppc64le/...)
Linux gc00 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.1 LTS
Release:	18.04
Codename:	bionic
- For GPU related issues:
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+

lsmod | grep nv_peer_mem
nv_peer_mem            16384  0
ib_core               323584  12 rdma_cm,ib_ipoib,mlx4_ib,nv_peer_mem,iw_cm,ib_iser,ib_umad,rdma_ucm,ib_uverbs,mlx5_ib,ib_cm,ib_ucm
nvidia              33988608  103 nvidia_uvm,nv_peer_mem,gdrdrv,nvidia_modeset

### Additional information (depending on the issue)
- ./contrib/configure-release --prefix=/usr/local/ucx --enable-mt --with-dm --with-cuda=/usr/local/cuda-11.2
configure output:
configure: =========================================================
configure: UCX build configuration:
configure:    Build prefix:  /usr/local/ucx
configure: Preprocessor flags:  -DCPU_FLAGS="" -I${abs_top_srcdir}/src -I${abs_top_builddir} -I${abs_top_builddir}/src
configure:     C compiler:  gcc -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -Wno-pointer-sign -Werror-implicit-function-declaration -Wno-format-zero-length -Wnested-externs -Wshadow -Werror=declaration-after-statement
configure:    C++ compiler:  g++ -O3 -g -Wall -Werror -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch
configure:    Multi-thread:  enabled
configure:     MPI tests:  disabled
configure:   Devel headers:  no
configure:      Bindings:  < >
configure:    UCT modules:  < cuda ib rdmacm cma knem >
configure:    CUDA modules:  < gdrcopy >
configure:    ROCM modules:  < >
configure:     IB modules:  < >
configure:    UCM modules:  < cuda >
configure:    Perf modules:  < cuda >
configure: =========================================================

**Comment by brminich** [comment]
@Seth5141  ucp_put/get_nbx operations do not support GPU memory

**Comment by shamisp** [comment]
@bureddy maybe you can comment when the constrain will be addressed as well propose a strategy to workaround this ? thanks

**Comment by Seth5141** [comment]
@brminich Thanks for clarifying that for me.

A couple follow up questions:
1. Which API families do support GPU memory? Is it just the tag and stream send/receive APIs? Or are there others?
2. I ran a little experiment where I mapped the local memory into the cpu process space with gdrcopy and passed the gdrcopy-mapped pointer into ucp_put_nbx. I had some success with that experiment before hitting another error. Do you know if that is a supported use-case for this API?


Thanks,

Seth

**Comment by Seth5141** [comment]
Hi all, I apologize for the confusion. I was going through an old e-mail thread with @yosefe and it appears that PUT/GET support is available in 1.10.0 with the UCX_PROTO_ENABLE environment variable. When I set that, I am able to get the behavior I need.
